### PR TITLE
feat(Protocol-state-filtering-mode) - Add support for protocol state filtering for VNI prototype

### PR DESCRIPTION
**Pull request by deepaksibm**
Resources:

- [x] VNI
- [x] Baremetal server
- [x] Baremetal server network attachment
- [x] Instance
- [x] Instance network attachment
- [x] Instance template
- [x] Share
- [x] Share mount target

Datasource:

- [x] VNI Collection
- [x] VNI

### CODE REVIEW

File: ibm/service/vpc/resource_ibm_is_share.go [hunk 1]
```diff
@@ -181,6 +181,13 @@ func ResourceIbmIsShare() *schema.Resource {
 										Computed:    true,
 										Description: "If `true`:- The VPC infrastructure performs any needed NAT operations.- `floating_ips` must not have more than one floating IP.If `false`:- Packets are passed unchanged to/from the network interface,  allowing the workload to perform any needed NAT operations.- `allow_ip_spoofing` must be `false`.- If the virtual network interface is attached:  - The target `resource_type` must be `bare_metal_server_network_attachment`.  - The target `interface_type` must not be `hipersocket`.",
 									},
+									"protocol_state_filtering_mode": {
```
Comment (ujjwal-ibm) [inline ibm/service/vpc/resource_ibm_is_share.go:184]: missing test cases and potentially data source changes, and test file for data source

File: ibm/service/vpc/resource_ibm_is_bare_metal_server.go [hunk 2]
```diff
@@ -569,6 +569,13 @@ func ResourceIBMIsBareMetalServer() *schema.Resource {
 										ValidateFunc:  validate.InvokeValidator("ibm_is_virtual_network_interface", "vni_name"),
 										Description:   "The name for this virtual network interface. The name is unique across all virtual network interfaces in the VPC.",
 									},
+									"protocol_state_filtering_mode": {
```
Comment (ujjwal-ibm) [inline ibm/service/vpc/resource_ibm_is_bare_metal_server.go:572]: missing test case, data source and data source test changes

File: ibm/service/vpc/resource_ibm_is_bare_metal_server_network_attachment.go [hunk 3]
```diff
@@ -172,6 +172,13 @@ func ResourceIBMIsBareMetalServerNetworkAttachment() *schema.Resource {
 							ValidateFunc:  validate.InvokeValidator("ibm_is_virtual_network_interface", "vni_name"),
 							Description:   "The name for this virtual network interface. The name is unique across all virtual network interfaces in the VPC.",
 						},
+						"protocol_state_filtering_mode": {
```
Comment (ujjwal-ibm) [inline ibm/service/vpc/resource_ibm_is_bare_metal_server_network_attachment.go:175]: missing test case, data source and data source test changes

File: ibm/service/vpc/resource_ibm_is_instance.go [hunk 4]
```diff
@@ -656,6 +656,13 @@ func ResourceIBMISInstance() *schema.Resource {
 										ValidateFunc:  validate.InvokeValidator("ibm_is_virtual_network_interface", "vni_name"),
 										Description:   "The name for this virtual network interface. The name is unique across all virtual network interfaces in the VPC.",
 									},
+									"protocol_state_filtering_mode": {
```
Comment (ujjwal-ibm) [inline ibm/service/vpc/resource_ibm_is_instance.go:659]: missing  data source and data source test changes

File: website/docs/r/is_bare_metal_server.markdown [hunk 5]
```diff
@@ -161,6 +161,12 @@ Review the argument references that you can specify for your resource.
       - `reserved_ip` - (Required, String) The unique identifier for this reserved IP.
       - `name` - (Required, String) The name for this reserved IP. The name is unique across all reserved IPs in a subnet.
       - `resource_type` - (Computed, String) The resource type.
+    - `protocol_state_filtering_mode` - (Optional, String) The protocol state filtering mode to use for this virtual network interface.
```
Comment (ujjwal-ibm) [inline website/docs/r/is_bare_metal_server.markdown:164]: missing data source documentation

**Review by ujjwal-ibm — CHANGES_REQUESTED** [review]

File: ibm/service/vpc/resource_ibm_is_share.go (hunk 1 above)
Comment (deepaksibm) [reply ibm/service/vpc/resource_ibm_is_share.go:184]: only the VNI and instance template API response contains protocol_state_filtering_mode. So only the data sources for VNI and VNI collection and template has the changes. 
Test cases are added

File: ibm/service/vpc/resource_ibm_is_bare_metal_server.go (hunk 2 above)
Comment (deepaksibm) [reply ibm/service/vpc/resource_ibm_is_bare_metal_server.go:572]: Test case added. Datasource changes not applicable as API response does not contain VNI details.

File: ibm/service/vpc/resource_ibm_is_instance.go (hunk 4 above)
Comment (deepaksibm) [reply ibm/service/vpc/resource_ibm_is_instance.go:659]: only the VNI and instance template API response contains protocol_state_filtering_mode. So only the data sources for VNI and VNI collection and template has the changes.
Test cases are added

File: website/docs/r/is_bare_metal_server.markdown (hunk 5 above)
Comment (deepaksibm) [reply website/docs/r/is_bare_metal_server.markdown:164]: only the VNI and instance template API response contains protocol_state_filtering_mode. So only the data sources for VNI and VNI collection and template has the changes.
Test cases are added

File: ibm/service/vpc/resource_ibm_is_bare_metal_server_network_attachment.go (hunk 3 above)
Comment (deepaksibm) [reply ibm/service/vpc/resource_ibm_is_bare_metal_server_network_attachment.go:175]: only the VNI and instance template API response contains protocol_state_filtering_mode. So only the data sources for VNI and VNI collection and template has the changes.
Test cases are added

**Comment by deepaksibm** [comment]
upstream merged and available 1.67.0

**Pull request closed without merging**